### ISSUE TITLE
Windows - Invalid repository

### ISSUE THREAD

**Issue by ijlyttle** [ropensci/git2r#329]
This started with #327, but I suspect that this is a separate issue because it occurs before `cred_ssh_key()` can ask me for my passphrase.

I am using dec6ef9 on Windows 10

```
> library("git2r")
> repo <- repository(".")
> push(repo, credentials = cred_ssh_key())
Error in head.git_repository(object) : 
  Error in 'git2r_repository_head': Invalid repository
```

**Comment by stewid** [comment]
Please try `master` branch again

**Comment by ijlyttle** [comment]
Using master branch, c8d86cc, I'm afraid I'm still getting the same thing.

**Comment by ijlyttle** [comment]
Yes - exactly the same error.

**Comment by stewid** [comment]
If you run
```
library("git2r")
repo <- repository(".")
.Call(git2r:::git2r_repository_can_open, repo$path)
```
does it return `TRUE` or `FALSE`?

**Comment by ijlyttle** [comment]
`TRUE`

**Comment by stewid** [comment]
I have still not been able to reproduce the issue. However, I ran the tests with valgrind (https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Using-valgrind) enabled and found some problems that I have fixed. I have also updated the bundled libgit2 to the latest commit.

Could you please try the master branch again?

**Comment by ijlyttle** [comment]
I'm afraid I get the same responses as before for `git2r::push()`, `git2r::repository_head()` and `.Call(git2r:::git2r_repository_can_open, repo$path)`. 

Using 17ce2e1eaca2b6860e8ac2e4b5ac89fe2f5aff96.

**Comment by jennybc** [comment]
I am getting exactly this message and behaviour with @maurolepore, also on Windows, installed from e2b3602.

**Comment by jeroen** [comment]
@ijlyttle @maurolepore could you test if this problem has been resolved in the windows branch?

```r
devtools::install_github("ropensci/git2r@windows")
```

**Comment by maurolepore** [comment]
I've just installed @windows (`75d23d0`) and the problem persists.  This output is the first line run on a new package (crated with `usethis::use_package()`):

```R
> use_git()
✔ Initialising Git repo
✔ Adding '.Rhistory', '.RData' to './.gitignore'
✔ Adding files and committing
Error in structure(.Call(git2r_status_list, lookup_repository(repo), staged,  : 
  Error in 'git2r_status_list': Invalid repository
```
```R
> traceback()
4: structure(.Call(git2r_status_list, lookup_repository(repo), staged, 
       unstaged, untracked, all_untracked, ignored), class = "git_status")
3: git2r::status(r)
2: unlist(git2r::status(r))
1: use_git()
```

**Comment by jeroen** [comment]
I cannot reproduce that problem either. What is the dir of the repository? It doesn't have spaces or so right?

**Comment by jennybc** [comment]
I think it's best to not involve usethis.

I'm posting from my Windows 10 VM. I also can't reproduce the problem.

``` r
library(git2r)

path <- "~/abcde"
dir.create(path)
r <- init(path)
r
#> Local:    C:/Users/JennyVM/Documents/abcde/
#> Head:     nothing commited (yet)

cat("hi", file = file.path(path, "foo.txt"))
status(r)
#> Untracked files:
#>  Untracked:  foo.txt

packageVersion("git2r")
#> [1] '0.21.0'
```

Created on 2018-05-24 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by maurolepore** [comment]
Above it was C:\Users\LeporeM\Documents\rm. It also fails with a repo at C:\Users\LeporeM\Dropbox\git_repos\rm2. 

If I start a new session, it seems that git has been initialized but no master branch created. And ` git2r::status()` fails.

```R
> git2r::status()
Error in structure(.Call(git2r_status_list, lookup_repository(repo), staged,  : 
  Error in 'git2r_status_list': Invalid repository
```

**Comment by jeroen** [comment]
There must be some local configuration issue in the mix. Does libgit2 have config files? Or perhaps it's trying and failing to read your keys or so?

**Comment by ijlyttle** [comment]
Sorry I'm late to this - for me, the original issue is still there: 

```
> library("git2r")
> repo <- repository(".")
> repository_head(repo)
```
```
Error in repository_head(repo) : 
  Error in 'git2r_repository_head': Invalid repository
```

This works OK for me:

```
> path <- "~/abcde"
> dir.create(path)
> r <- init(path)
```

However...

```
> repository_head(path)
```
```
Error in repository_head(path) : 
  Error in 'git2r_repository_head': Invalid repository
```
<details>
<summary> Session info</summary>
<pre>
Session info -----------------------------------------------------------------------------
 setting  value                       
 version  R version 3.4.3 (2017-11-30)
 system   x86_64, mingw32             
 ui       RStudio (1.1.414)           
 language (EN)                        
 collate  English_United States.1252  
 tz       America/Chicago             
 date     2018-05-25                  

Packages ---------------------------------------------------------------------------------
 package    * version     date       source                             
 base       * 3.4.3       2017-12-06 local                              
 compiler     3.4.3       2017-12-06 local                              
 datasets   * 3.4.3       2017-12-06 local                              
 devtools   * 1.13.5      2018-02-18 CRAN (R 3.4.3)                     
 digest       0.6.15      2018-01-28 CRAN (R 3.4.3)                     
 git2r      * 0.21.0.9002 2018-05-25 Github (ropensci/git2r@75d23d0)    
 graphics   * 3.4.3       2017-12-06 local                              
 grDevices  * 3.4.3       2017-12-06 local                              
 grid         3.4.3       2017-12-06 local                              
 jsonlite     1.5         2017-06-01 CRAN (R 3.4.3)                     
 lattice      0.20-35     2017-03-25 CRAN (R 3.4.3)                     
 Matrix       1.2-12      2017-11-20 CRAN (R 3.4.3)                     
 memoise      1.1.0       2017-04-21 CRAN (R 3.4.3)                     
 methods    * 3.4.3       2017-12-06 local                              
 Rcpp         0.12.16     2018-03-13 CRAN (R 3.4.4)                     
 reticulate   1.6.0.9002  2018-04-29 Github (rstudio/reticulate@7025d11)
 stats      * 3.4.3       2017-12-06 local                              
 tools        3.4.3       2017-12-06 local                              
 utils      * 3.4.3       2017-12-06 local                              
 withr        2.1.1       2017-12-19 CRAN (R 3.4.3)                     
 yaml         2.1.18      2018-03-08 CRAN (R 3.4.3) 
</pre>
</details>

**Comment by jeroen** [comment]
> Would keys matter for running status on a local repo, though?

I would expect not, but there must be something that's different in their setup that trips over libgit2.

**Comment by jennybc** [comment]
I've rebuilt from affb836f222b96e7db06727bf323bfc9499c3b90 on Windows and confirm that all is still well for me. I note there are few little notes/warnings from R CMD check that probably need to be dealt with before CRAN. But obviously this other problem is a much bigger blocker.

**Comment by jeroen** [comment]
@stewid and me added some more debugging output. @ijlyttle @maurolepore could you install from source one more time, and see if it gives you a better error message?

```r
devtools::install_github("ropensci/git2r")
```

As always make sure to install in a clean R session. If the installation ends with some line *could not replace git2r.dll* or something like that, the update was not successful.

**Comment by stewid** [comment]
I have added a print debugging line in the C code (in the function that opens a repository) to see if the repository object is invalid (for example a memory problem in the C code) or if the code is unable to open the repository. Does it print `The repo argument is unexpectedly invalid` before the error occurs?

`Rprintf` added: https://github.com/ropensci/git2r/blob/master/src/git2r_repository.c#L45

**Comment by maurolepore** [comment]
Now I can successfully run all examples in `?add()`.

``` r
packageVersion("git2r")
#> [1] ‘0.21.0.9002’
```

But I fail to push with git2r, although I can push from the terminal.

``` r
library(git2r)


path_repo <- "C:/Users/LeporeM/Documents/abc1"
repo <- init(path_repo)
config(repo, user.name="maurolepore", user.email="maurolepore@gmail.com")

file.create("d.txt")
#> [1] TRUE
writeLines("Test push", file.path(path_repo, "d.txt"))

add(repo, "d.txt")
commit(repo, "First commit message")
#> [cdd7102] 2018-05-25: First commit message

push(repo, "origin", "refs/heads/master")
#> Error in push(repo, "origin", "refs/heads/master"): Error in 'git2r_push': error authenticating: failed connecting agent
```

Created on 2018-05-25 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by ijlyttle** [comment]
I have reinstalled with no errors, and I get the same responses as [above](https://github.com/ropensci/git2r/issues/329#issuecomment-391942747).

I get no additional error message before the error occurs.

<details>
<summary> Session info</summary>
<pre>
 session_info()
Session info ------------------------------------------------------------------------------------------
 setting  value                       
 version  R version 3.4.3 (2017-11-30)
 system   x86_64, mingw32             
 ui       RStudio (1.1.414)           
 language (EN)                        
 collate  English_United States.1252  
 tz       America/Chicago             
 date     2018-05-25                  

Packages ----------------------------------------------------------------------------------------------
 package    * version     date       source                             
 base       * 3.4.3       2017-12-06 local                              
 compiler     3.4.3       2017-12-06 local                              
 curl         3.1         2017-12-12 CRAN (R 3.4.3)                     
 datasets   * 3.4.3       2017-12-06 local                              
 devtools   * 1.13.5      2018-02-18 CRAN (R 3.4.3)                     
 digest       0.6.15      2018-01-28 CRAN (R 3.4.3)                     
 git2r      * 0.21.0.9002 2018-05-25 Github (ropensci/git2r@70b9a12)    
 graphics   * 3.4.3       2017-12-06 local                              
 grDevices  * 3.4.3       2017-12-06 local                              
 grid         3.4.3       2017-12-06 local                              
 httr         1.3.1       2017-08-20 CRAN (R 3.4.3)                     
 jsonlite     1.5         2017-06-01 CRAN (R 3.4.3)                     
 lattice      0.20-35     2017-03-25 CRAN (R 3.4.3)                     
 Matrix       1.2-12      2017-11-20 CRAN (R 3.4.3)                     
 memoise      1.1.0       2017-04-21 CRAN (R 3.4.3)                     
 methods    * 3.4.3       2017-12-06 local                              
 R6           2.2.2       2017-06-17 CRAN (R 3.4.3)                     
 Rcpp         0.12.16     2018-03-13 CRAN (R 3.4.4)                     
 reticulate   1.6.0.9002  2018-04-29 Github (rstudio/reticulate@7025d11)
 rstudioapi   0.7         2017-09-07 CRAN (R 3.4.3)                     
 stats      * 3.4.3       2017-12-06 local                              
 tools        3.4.3       2017-12-06 local                              
 utils      * 3.4.3       2017-12-06 local                              
 withr        2.1.1       2017-12-19 CRAN (R 3.4.3)                     
 yaml         2.1.18      2018-03-08 CRAN (R 3.4.3) 
</pre>
</details>

**Comment by maurolepore** [comment]
Here is something interesting about my git configuration. My details are found if requested from RStudio terminal but not if requested from git bash.

![image](https://user-images.githubusercontent.com/5856545/40566333-29502592-603e-11e8-9ab1-52db62955d9f.png)

**Comment by stewid** [comment]
I found a problem when copying the repo object internally that could possibly cause this. @ijlyttle does it work now?

**Comment by ijlyttle** [comment]
Just a speculation, @maurolepore, but I wonder if the difference in behavior could be caused by RStudio and bash having different definitions of the git home directory. To be clear, I don't know if they are, in fact,  different or not. We found something along these lines with ssh: https://github.com/ropensci/git2r/issues/314

**Comment by ijlyttle** [comment]
Still the same responses as [above](https://github.com/ropensci/git2r/issues/329#issuecomment-391942747).

I get no additional error message before the error occurs.

<details>
<summary> Session info</summary>
<pre>
Session info ------------------------------------------------------------------------------
 setting  value                       
 version  R version 3.4.3 (2017-11-30)
 system   x86_64, mingw32             
 ui       RStudio (1.1.414)           
 language (EN)                        
 collate  English_United States.1252  
 tz       America/Chicago             
 date     2018-05-27                  

Packages ----------------------------------------------------------------------------------
 package    * version     date       source                             
 base       * 3.4.3       2017-12-06 local                              
 compiler     3.4.3       2017-12-06 local                              
 datasets   * 3.4.3       2017-12-06 local                              
 devtools   * 1.13.5      2018-02-18 CRAN (R 3.4.3)                     
 digest       0.6.15      2018-01-28 CRAN (R 3.4.3)                     
 git2r      * 0.21.0.9002 2018-05-26 Github (ropensci/git2r@461b9be)    
 graphics   * 3.4.3       2017-12-06 local                              
 grDevices  * 3.4.3       2017-12-06 local                              
 grid         3.4.3       2017-12-06 local                              
 jsonlite     1.5         2017-06-01 CRAN (R 3.4.3)                     
 lattice      0.20-35     2017-03-25 CRAN (R 3.4.3)                     
 Matrix       1.2-12      2017-11-20 CRAN (R 3.4.3)                     
 memoise      1.1.0       2017-04-21 CRAN (R 3.4.3)                     
 methods    * 3.4.3       2017-12-06 local                              
 Rcpp         0.12.16     2018-03-13 CRAN (R 3.4.4)                     
 reticulate   1.6.0.9002  2018-04-29 Github (rstudio/reticulate@7025d11)
 stats      * 3.4.3       2017-12-06 local                              
 tools        3.4.3       2017-12-06 local                              
 utils      * 3.4.3       2017-12-06 local                              
 withr        2.1.1       2017-12-19 CRAN (R 3.4.3)                     
 yaml         2.1.18      2018-03-08 CRAN (R 3.4.3)  
</pre>
</details>

**Comment by jennybc** [comment]
@ijlyttle on the machine in question, would you be able to give someone remote control briefly in this way:

https://support.zoom.us/hc/en-us/articles/201362673-Request-or-Give-Remote-Control

I'm trying to lay the groundwork for getting this solved. I probably am not the one to fix it, but I can at least do some grunt work. I assume your Troubleshooter might also ask that you install a debugger.

**Comment by jennybc** [comment]
@ijlyttle Me again. Can you downgrade to, say, the CRAN version of git2r and do some basic git2r operations?   This would help clarify if something is generally borked on your machine or if it's specific to the dev version of git2r.

**Comment by jeroen** [comment]
@ijlyttle could you somehow rule out that `.gitconfig` is involved? E.g. temporarily rename it to some other file.

**Comment by ijlyttle** [comment]
Hi @jennybc - I am happy to help however I can, but I will be traveling from tomorrow until next Thursday (but the computer in question will not be).

I tried using the CRAN version, but I cannot get that to work either - starting to suspect that the computer is borked.

Hi @jeroen - I tried also renaming `.gitconfig`, but to no effect.

**Comment by maelle** [comment]
:wave: my dear Windows friends! Let's find 🐛 together 💃 

I have tested the code below

```r
library("git2r")
repo_path <- "therepo"
fs::dir_create(repo_path)
fs::file_create(file.path(repo_path,"somefile.txt"))
r <- git2r::init(repo_path)
paths <- unlist(git2r::status(r))
git2r::add(r, paths)
git2r::commit(r, "1st commit")
git2r::head(r)
git2r::remote_add(r, "origin", "git@github.com:maelle/git2rtests")
git2r::push(r, "origin", "refs/heads/master")
fs::dir_delete(repo_path)
```

With the CRAN version, it worked until `git2r::push`, with the error

```r
Error in .local(object, ...) : 
  Error in 'git2r_push': failed to start SSH session: Unable to exchange encryption keys
```

That said when I tried pushing with `git2r` a while ago (when @ijlyttle opened the issue), I think I had a different error, but at pushing too. I can push from the terminal.

Now with the dev version I can reproduce the error 🎊 😭 

``` r
library("git2r")
repo_path <- "therepo"
fs::dir_create(repo_path)
fs::file_create(file.path(repo_path,"somefile.txt"))
r <- git2r::init(repo_path)
git2r::status(r)
#> Error in structure(.Call(git2r_status_list, lookup_repository(repo), staged, : Error in 'git2r_status_list': Invalid repository
```

Created on 2018-06-02 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

Renaming `.gitconfig` changed nothing.

I am happy to keep testing stuff if useful, and @jeroen I can do a call on Monday if easier.

**Comment by jennybc** [comment]
Thanks @maelle! It will be wonderful to get to the bottom of this and it's going to be a great help to have more people who can reproduce it.

Next time you return, can you provide details on R's notion of your home directory and, in that context, where your user-level gitconfig lives. And your SSH keys? I don't really think this is the problem, but I'm trying to have an open mind. And I have a sideline interest in clarifying these matters as well.

This section has some relevant words and code, focusing SSH:
http://happygitwithr.com/ssh-keys.html#ssh-troubleshooting

**Comment by maelle** [comment]
🕵️‍♀️  -- these things are relevant for my trying to work with the CRAN version right? The dev version seems to block me at an earlier stage.

# R's notion of my home directory
``` r
normalizePath("~")
#> [1] "C:\\Users\\Maelle\\Documents"
as.list(Sys.getenv(
  c("HOME", "USERPROFILE")
))
#> $HOME
#> [1] "C:/Users/Maelle/Documents"
#> 
#> $USERPROFILE
#> [1] "C:\\Users\\Maelle"
```

Created on 2018-06-02 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

# Where my user-level gitconfig lives
``` r
list.files(
  Sys.getenv("USERPROFILE"),
  pattern = "ssh|git",
  include.dirs = TRUE,
  all.files = TRUE
)
#> [1] ".gitconfig" ".ssh"
```

Created on 2018-06-02 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

## Debugging in the book

I had actually spent some time reading Happy/Sad Git and GitHub with R :wink: and already have a symlink

![image](https://user-images.githubusercontent.com/8360597/40878834-33e28914-6697-11e8-910e-5ba152953ff7.png)

I could re-install the CRAN version to try "telling git2r explicitly where to find my public and private key"? Would it help to see if this fails?

**Comment by ijlyttle** [comment]
Just to add a quick two cents - my last message where I thought my Windows computer was borked using the CRAN version - @maelle's experience is entirely consistent with mine (except she had the 💪to document everything).

**Comment by jeroen** [comment]
I think the problem here is that when you try to install `git2r` using devtools, the installation may be corrupted because devtools itself uses git2r, and windows fails to update the dll file. Did you see a line like this when installing?
```
Warning in file.copy(files, dest, overwrite = TRUE) :
  problem copying .\git2r.dll to C:\Users\Jeroen\Documents\R\win-library\3.4\git2r\libs\x64\git2r.dll: Permission denied
```


<img width="1045" alt="screen shot 2018-06-06 at 5 40 20 pm" src="https://user-images.githubusercontent.com/216319/41049501-5b7d5702-69b1-11e8-9c50-5c7677e6923b.png">

Could you guys and girls try installing from the preocmpiled binary package from appveyor? To do so, click on the appveyor badge in the git2r readme, go to the latest build and click on __Artifacts__ in the top right corner, and download the binary package zip file `git2r_0.21.0.9002.zip`.

<img width="1037" alt="screen shot 2018-06-06 at 5 51 34 pm" src="https://user-images.githubusercontent.com/216319/41049993-74e86f96-69b2-11e8-91f7-e734ada60f79.png">

You can install this zip file using `install.packages("git2r_0.21.0.9002.zip", repos = NULL)` or in the GUI and rstudio there is a button for this as well. I'm interested if the problem still persists then.

**Comment by jimhester** [comment]
To avoid that issue you can install from github using remotes, which does not depend on git2r.

**Comment by maelle** [comment]
Installed with `remotes` (thanks @jimhester!)

``` r
library("git2r")
repo_path <- "therepo"
fs::dir_create(repo_path)
fs::file_create(file.path(repo_path,"somefile.txt"))
r <- git2r::init(repo_path)
paths <- unlist(git2r::status(r))
git2r::add(r, paths)
git2r::commit(r, "1st commit")
#> [44ad0e1] 2018-06-06: 1st commit
git2r::head(r)
#> Warning: 'head.git_repository' is deprecated.
#> Use 'repository_head' instead.
#> See help("Deprecated")
#> [44ad0e] (Local) (HEAD) master
git2r::remote_add(r, "origin", "git@github.com:maelle/git2rtests")
git2r::push(r, "origin", "refs/heads/master")
#> Error in git2r::push(r, "origin", "refs/heads/master"): Error in 'git2r_push': error authenticating: failed connecting agent
fs::dir_delete(repo_path)
```

Created on 2018-06-06 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

<details>

<summary>Session info</summary>

``` r
devtools::session_info()
#> Session info -------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.5.0 (2018-04-23)
#>  system   x86_64, mingw32             
#>  ui       RTerm                       
#>  language (EN)                        
#>  collate  English_United States.1252  
#>  tz       Europe/Paris                
#>  date     2018-06-06
#> Packages -----------------------------------------------------------------
#>  package   * version     date       source                         
#>  backports   1.1.2       2017-12-13 CRAN (R 3.5.0)                 
#>  base      * 3.5.0       2018-04-23 local                          
#>  compiler    3.5.0       2018-04-23 local                          
#>  datasets  * 3.5.0       2018-04-23 local                          
#>  devtools    1.13.5      2018-02-18 CRAN (R 3.5.0)                 
#>  digest      0.6.15      2018-01-28 CRAN (R 3.5.0)                 
#>  evaluate    0.10.1      2017-06-24 CRAN (R 3.5.0)                 
#>  fs          1.2.2       2018-03-21 CRAN (R 3.5.0)                 
#>  git2r     * 0.21.0.9002 2018-06-06 Github (ropensci/git2r@73db600)
#>  graphics  * 3.5.0       2018-04-23 local                          
#>  grDevices * 3.5.0       2018-04-23 local                          
#>  htmltools   0.3.6       2017-04-28 CRAN (R 3.5.0)                 
#>  knitr       1.20        2018-02-20 CRAN (R 3.5.0)                 
#>  magrittr    1.5         2014-11-22 CRAN (R 3.5.0)                 
#>  memoise     1.1.0       2017-04-21 CRAN (R 3.5.0)                 
#>  methods   * 3.5.0       2018-04-23 local                          
#>  Rcpp        0.12.17     2018-05-18 CRAN (R 3.5.0)                 
#>  rmarkdown   1.9         2018-03-01 CRAN (R 3.4.4)                 
#>  rprojroot   1.3-2       2018-01-03 CRAN (R 3.4.3)                 
#>  stats     * 3.5.0       2018-04-23 local                          
#>  stringi     1.2.2       2018-05-02 CRAN (R 3.5.0)                 
#>  stringr     1.3.1       2018-05-10 CRAN (R 3.5.0)                 
#>  tools       3.5.0       2018-04-23 local                          
#>  utils     * 3.5.0       2018-04-23 local                          
#>  withr       2.1.2       2018-03-15 CRAN (R 3.4.4)                 
#>  yaml        2.1.19      2018-05-01 CRAN (R 3.5.0)
```

</details>

**Comment by jeroen** [comment]
@maelle thanks! So the password prompt is still not working well, but definitely no regressions from the latest cran version.

**Comment by jennybc** [comment]
@maelle Is your SSH key passphrase protected?

**Comment by maelle** [comment]
@jennybc no not at the moment

**Comment by jennybc** [comment]
Interesting. Have you ever been able to push using SSH protocol on Windows with git2r? I've always suspected that perhaps this has never worked? And yet I feel like maybe @ijlyttle has reported success in the past? Apologies to the universe for muddying the water re: this particular issue thread with someone separate.

This seems relevant, in terms of which ssh agent libgit2 is ready to work with on Windows:
https://github.com/nodegit/nodegit/issues/1040

**Comment by maelle** [comment]
Now that I added a passphrase ☺️ I get the same error message. So I can keep experimenting for y'all, since I didn't break things any further on my system.

The reason why I tried getting SSH protocol working with `git2r` back in February was because @ijlyttle reported success. Then after trying for what seemed a long time, I asked him and he told me it no longer worked for him.

**Comment by maelle** [comment]
I can experiment with and without passphrase if it helps! But it seems that's not the part where problems happen anyway.

**Comment by jennybc** [comment]
To be clear, having a passphrase will make things *more* difficult. I wondered if that was the problem.

**Comment by stewid** [comment]
Thanks @maelle 

> Have you ever been able to push using SSH protocol on Windows with git2r?

Yes (https://github.com/stewid/test-07)

**Comment by stewid** [comment]
@maelle does it work if you specify the path to the SSH key?

```
cred <- git2r::cred_ssh_key("path/to/id_rsa.pub", "path/to/id_rsa")
git2r::push(r, "origin", "refs/heads/master", credentials = cred)
```

**Comment by maelle** [comment]
It does work indeed! 🎉 https://github.com/maelle/git2rtests

I was asked for my passphrase, and then it worked.

**Comment by jennybc** [comment]
Are we starting to conclude this?

  * Previous problems were the result of borked installations of git2r. Or ... perhaps ... previous problems were legit, have been actively resolved, but initial attempts to verify this were derailed by corrupt re-installations of dev git2r, due to devtools' usage of git2r.
  * A separate problem exists, in which SSH keys aren't being found on Windows when perhaps they should be.

**Comment by maelle** [comment]
@jeroen suggested copying the `.ssh` folder to Documents (where there only was a symlink), and it didn't help, I still need to explicitely indicate where the key is.

**Comment by stewid** [comment]
Currently, git2r is looking for the default ssh keys in
```
file.path(Sys.getenv("USERPROFILE"), ".ssh")
```

@maelle , are your ssh keys in
```
file.path(Sys.getenv("HOMEDRIVE", Sys.getenv("HOMEPATH", ".ssh")))
```

**Comment by ijlyttle** [comment]
Hi all,

As usual, I am late to the party 🎉

Just to report that after using **remotes** to install **git2r**, all of the actions that had produced errors **have stopped producing errors** 💃

@maelle: I don't know if this can be useful or not (I see that you have solved the problem), but this was my workaround for SSH keys: https://ijlyttle.github.io/ghentr/articles/workarounds.html#workaround. These ideas have already been touched-on by @stewid and @jennybc, but thought I would add my two cents. From what I remember, this was addressed in #314.

For me, I am happy to close the issue, but I know there are six others who may have an opinion.

**Comment by maelle** [comment]
Thanks @ijlyttle, it looks like the symlink I already have (maybe I even got the solution from that page?). 

@stewid 

``` r
file.path(Sys.getenv("HOMEDRIVE"), Sys.getenv("HOMEPATH"), ".ssh")
#> [1] "C:/\\Users\\Maelle/.ssh"
```

Created on 2018-06-08 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by maelle** [comment]
Currently my .ssh resides under both Maelle and Maelle/Documents.

**Comment by stewid** [comment]
Thanks @ijlyttle . Glad to hear that it works. I'm also happy to close the issue.

> A separate problem exists, in which SSH keys aren't being automatically found on Windows, when perhaps they should be.

I open a new issue for this.

Thanks to you all!
